### PR TITLE
KMTF fixes on propagation coefficients (CMSSW_11_3_X)

### PR DESCRIPTION
**Pull request by Tyler-Lam**
#### PR description:

New version of KMTF algorithm as discussed in https://indico.cern.ch/event/1012620/contributions/4250430/attachments/2200192/3721090/Run3_KMTF.pdf

#### PR validation:

Nominal tests with just KMTF algorithm. No core changes, just incremental changes to KMTF

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

@rekovic The lookup tables located in L1Trigger/L1TMuon/data/bmtf_luts/kalmanLUTs.root are usually in the https://github.com/cms-l1t-offline/L1Trigger-L1TMuon.git repository.  They have to be moved but I do not have access
@dinyar please check that we didn't revert your changes

### CODE REVIEW

**Comment by silviodonato** [comment]
please test
kind reminder @cms-sw/l1-l2

**Comment by dinyar** [comment]
Hi @Tyler-Lam,

Thanks for the ping, looks good to me.

Cheers,
Dinyar

**Comment by cecilecaillol** [comment]
+l1

File: L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanAlgo.cc [hunk 1]
```diff
@@ -353,25 +358,41 @@ void L1TMuonBarrelKalmanAlgo::propagate(L1MuKBMTrack& track) {
...
   //phi propagation
-  int phi11 = fp_product(aPhi_[step - 1], K, 10);
-  int phi12 = fp_product(-bPhi_[step - 1], phiB, 10);
+  ap_fixed<BITSCURV, BITSCURV> phi11 = ap_fixed<BITSPARAM + 1, 2>(aPhi_[step - 1]) * ap_fixed<BITSCURV, BITSCURV>(K);
+  ap_fixed<BITSPHIB, BITSPHIB> phi12 =
+      ap_fixed<BITSPARAM + 1, 2>(-bPhi_[step - 1]) * ap_fixed<BITSPHIB, BITSPHIB>(phiB);
+
   if (verbose_) {
-    printf("phi prop = %d* %f = %d, %d* %f =%d\n", K, aPhi_[step - 1], phi11, phiB, -bPhi_[step - 1], phi12);
+    std::cout << "phi prop = " << K << "*" << ap_fixed<BITSPARAM + 1, 2>(aPhi_[step - 1]) << " = " << phi11 << ", "
```
Comment (silviodonato) [inline L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanAlgo.cc:366]: You should avoid using `std::cout`, please use https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger
(see https://github.com/cms-sw/cms-sw.github.io/issues/94).
Please fix it in all the file (18 times)

File: L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanAlgo.cc [hunk 2]
```diff
@@ -353,25 +358,41 @@ void L1TMuonBarrelKalmanAlgo::propagate(L1MuKBMTrack& track) {
...
-  int phiB11 = fp_product(aPhiB_[step - 1], K, 10);
-  int phiB12 = fp_product(bPhiB_[step - 1], phiB, 11);
-  int phiBNew = wrapAround(phiB11 + phiB12, 4096);
+  ap_fixed<BITSCURV, BITSCURV> phiB11 = ap_fixed<BITSPARAM, 1>(aPhiB_[step - 1]) * ap_fixed<BITSCURV, BITSCURV>(K);
+  ap_fixed<BITSPHIB + 1, BITSPHIB + 1> phiB12 =
+      ap_ufixed<BITSPARAM + 1, 1>(bPhiB_[step - 1]) * ap_fixed<BITSPHIB, BITSPHIB>(phiB);
+  int phiBNew = ap_fixed<13, 13>(phiB11 + phiB12);
   if (verbose_) {
-    printf("phiB prop = %d* %f = %d, %d* %f =%d\n", K, aPhiB_[step - 1], phiB11, phiB, bPhiB_[step - 1], phiB12);
+    std::cout << "phiB prop = " << K << "* " << ap_fixed<BITSPARAM + 1, 2>(aPhiB_[step - 1]) << " = " << phiB11 << ", "
```
Comment (silviodonato) [inline L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanAlgo.cc:377]: same here

File: L1Trigger/L1TMuonBarrel/interface/L1TMuonBarrelKalmanAlgo.h [hunk 3]
```diff
@@ -106,6 +106,47 @@ class L1TMuonBarrelKalmanAlgo {
...
+  static const int GAIN_0 = 18;
+  static const int GAIN_0INT = 7;
+  static const int GAIN_1 = 18;
+  static const int GAIN_1INT = 4;
+  static const int GAIN_4 = 18;
+  static const int GAIN_4INT = 4;
+  static const int GAIN_5 = 18;
+  static const int GAIN_5INT = 0;
+
+  static const int GAIN2_0 = 18;
```
Comment (silviodonato) [inline L1Trigger/L1TMuonBarrel/interface/L1TMuonBarrelKalmanAlgo.h:139]: please remove the comment lines

**Comment by silviodonato** [comment]
@Tyler-Lam @cms-sw/l1-l2
could you add a line saying explicitly that these differences are expected?

> DQMHistoTests: Total failures: 429

File: L1Trigger/L1TMuonBarrel/test/kalmanTools/parseGainsLUT.py [hunk 4]
```diff
@@ -138,100 +158,105 @@ def parse(fileio,f,fout,ele,bins,mini,maxi,zeroSup=False,constant=-1,absol=False
...
 
-parse(fileio,f,fout,0,512,0,64,True,-1,False,2)    
-parse(fileio,f,fout,4,512,0,16,True,-1,True,2)    
+parse(fileio,f,fout,0,512*2,0,64*2,True,-1,False,2,False)    
+parse(fileio,f,fout,4,512,0,16,True,-1,True,2,False)    
 
 parse2(fileio,f,fout,0,512,0,64,True,-1,False,1)    
-parse2(fileio,f,fout,1,512,-8,8,True,-1,False,1)    
+parse2(fileio,f,fout,1,512,-8,8,True,-1,True,1)    
+#parse2(fileio,f,fout,1,512,-8,8,True,-1,False,1)
```
Comment (rekovic) [inline L1Trigger/L1TMuonBarrel/test/kalmanTools/parseGainsLUT.py:254]: remove commented line

File: L1Trigger/L1TMuonBarrel/test/kalmanTools/parseGainsLUT.py [hunk 5]
```diff
@@ -51,36 +52,55 @@ def generate(tfile,name,outname,bins,mini,maxi,zeroSup=False,constant=-1,absol=F
...
 fileio=open("gainLUTs.h","w")
-fileio.write('#include "ap_fixed.h"\n')
-fileio.write('typedef ap_ufixed<9,6> updateLUT0_t;\n')
-fileio.write('typedef ap_ufixed<9,4> updateLUT4_t;\n')
-fileio.write('typedef ap_fixed<9,4> updateLUT1_t;\n')
-fileio.write('typedef ap_ufixed<9,8> updateLUT5_t;\n')
-fileio.write('typedef ap_ufixed<9,2> updateLUTV0_t;\n')
-fileio.write('typedef ap_ufixed<9,2> updateLUTV1_t;\n');
+fileio.write('#include common.h\n')
+#fileio.write('typedef ap_ufixed<9,6> updateLUT0_t;\n')
```
Comment (rekovic) [inline L1Trigger/L1TMuonBarrel/test/kalmanTools/parseGainsLUT.py:98]: remove commented lines

File: L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanAlgo.cc (hunk 1 above)
Comment (Tyler-Lam) [reply L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanAlgo.cc:366]: Would reverting to printf() statements be ok?  We use this format to make hardware patterns that can be very expensive, so the activation of the logger would give a lot of output for no reason.

**Comment by rekovic** [comment]
@Tyler-Lam 
We would like this backported to 11_2_X, so please address the comments.

**Comment by Tyler-Lam** [comment]
@rekovic I replied previously but forgot to tag @silviodonato, this emulator has to have large printouts for hardware patterns and the messagelogger output would interfere with this. I've removed the couts and reverted them to printf()

@silviodonato The emulator has been changed so we expect differences in the DQM distributions

**Comment by dinyar** [comment]
Hi @Tyler-Lam,

Out of interest, why are these printouts necessary for hardware patterns? Can't you just write them to a file? Reason I'm asking is that we'll probably eventually move our own pattern generators into CMSSW and I'm curious how others are tackling it.

Cheers,
Dinyar

**Comment by bachtis** [comment]
Hi @dinyar 
What we do is have our test benches receive standard I/O. We could have it written to file but there is no reason for this or not. I prefer to stay backwards compatible 
@rekovic , KMTF was always like this I do not understand why we cannot leave our verbose flag as it was and continue to use what we have implemented . Is this OK?

**Comment by silviodonato** [comment]
please test

**Comment by rekovic** [comment]
Ok to sign off for L1 after passing tests.

**Comment by silviodonato** [comment]
please test
the error was already in the IB

**Comment by bachtis** [comment]
@rekovic seems this is ready?

**Comment by panoskatsoulis** [comment]
Had a quick look at some of the comparison-plots as well and I think all look good - no unexpected changes in the WFs
I assume that minor changes in the PT distributions are because of the modified coefficients that lead to slightly different tracks and curvatures. (example here https://tinyurl.com/yzrwk4a5)

I think this is ready for the backport 
(we need to catch the deadline of the DQM so if also @rekovic and @silviodonato think that this is ready to go, 
please open the 11_2_X PR too)

**Comment by bachtis** [comment]
@panoskatsoulis as you see Tyler has opened the backport

**Comment by panoskatsoulis** [comment]
> @panoskatsoulis as you see Tyler has opened the backport

right my bad, mentioning this to the DQM
thx

**Comment by silviodonato** [comment]
@cms-sw/l1-l2 do you have any comments?

**Comment by cecilecaillol** [comment]
+l1

**Comment by qliphy** [comment]
+1

**Pull request merged**